src/components/new-safe/steps/Step4/logic/index.ts

```diff
@@ -0,0 +1,232 @@
+  })
+
+  return {
+    data,
+    from: wallet.address,
+    nonce: await provider.getTransactionCount(wallet.address),
+    to: proxyContract.getAddress(),
+    value: BigNumber.from(0),
+    startBlock: await provider.getBlockNumber(),
+  }
+}
+
+export type SafeCreationProps = {
+  owners: string[]
+  threshold: number
+  saltNonce: number
+}
+
+export const estimateSafeCreationGas = async (
+  chain: ChainInfo,
+  provider: JsonRpcProvider,
+  from: string,
+  safeParams: SafeCreationProps,
+): Promise<BigNumber> => {
+  const proxyFactoryContract = getProxyFactoryContractInstance(chain.chainId)
+  const encodedSafeCreationTx = encodeSafeCreationTx({ ...safeParams, chain })
+
+  return provider.estimateGas({
+    from: from,
+    to: proxyFactoryContract.getAddress(),
+    data: encodedSafeCreationTx,
+  })
+}
+
+export const pollSafeInfo = async (chainId: string, safeAddress: string): Promise<SafeInfo> => {
+  // exponential delay between attempts for around 4 min
+  return backOff(() => getSafeInfo(chainId, safeAddress), {
+    startingDelay: 750,
+    maxDelay: 20000,
+    numOfAttempts: 19,
+    retry: (e) => {
+      console.info('waiting for client-gateway to provide safe information', e)
+      return true
+    },
+  })
+}
+
+export const handleSafeCreationError = (error: EthersError) => {
+  logError(Errors._800, error.message)
+
+  if (isWalletRejection(error)) {
+    return SafeCreationStatus.WALLET_REJECTED
+  }
+
+  if (error.code === ErrorCode.TRANSACTION_REPLACED) {
+    if (error.reason === 'cancelled') {
+      return SafeCreationStatus.ERROR
+    } else {
+      return SafeCreationStatus.SUCCESS
+    }
+  }
+
+  return SafeCreationStatus.TIMEOUT
+}
+
+export const checkSafeCreationTx = async (
+  provider: JsonRpcProvider,
+  pendingTx: PendingSafeTx,
+  txHash: string,
+): Promise<SafeCreationStatus> => {
+  const TIMEOUT_TIME = 6.5 * 60 * 1000 // 6.5 minutes
+
+  try {
+    const receipt = await provider._waitForTransaction(txHash, 1, TIMEOUT_TIME, pendingTx)
+
+    if (didRevert(receipt)) {
+      return SafeCreationStatus.REVERTED
+    }
+
+    return SafeCreationStatus.SUCCESS
+  } catch (err) {
+    return handleSafeCreationError(err as EthersError)
+  }
+}
+
+export const getRedirect = (
+  chainId: string,
+  safeAddress: string,
+  redirectQuery?: string | string[],
+): UrlObject | string => {
+  const redirectUrl = Array.isArray(redirectQuery) ? redirectQuery[0] : redirectQuery
+  const chainPrefix = Object.keys(chains).find((prefix) => chains[prefix] === chainId)
+  const address = `${chainPrefix}:${safeAddress}`
+
+  // Should never happen in practice
+  if (!chainPrefix) return AppRoutes.index
+
+  // Go to the dashboard if no specific redirect is provided
+  if (!redirectUrl) {
+    return { pathname: AppRoutes.home, query: { safe: address, creation: 1 } }
+  }
+
+  // Otherwise, redirect to the provided URL (e.g. from a Safe App)
+
+  // Track the redirect to Safe App
+  if (redirectUrl.includes('apps')) {
```

Comment:
The only other route that could be affected is `/settings/safe-apps`. I used `AppRoutes.apps` for the comparison instead which should narrow it down.

Reply:
I originally copied this function from the old create-safe flow but to avoid code duplication I removed it now and added TODOs so we can refactor it at a later time.